framework/src/main/java/org/checkerframework/framework/util/dependenttypes/DependentTypesHelper.java

```diff
@@ -256,198 +272,86 @@ private void viewpointAdaptExecutable(
+        StringToJavaExpression stringToJavaExpr;
+        if (tree instanceof MethodInvocationTree) {
+            stringToJavaExpr =
+                    stringExpr ->
+                            StringToJavaExpression.atMethodInvocation(
+                                    stringExpr, (MethodInvocationTree) tree, factory.getChecker());
+        } else if (tree instanceof NewClassTree) {
+            stringToJavaExpr =
+                    stringExpr ->
+                            StringToJavaExpression.atConstructorInvocation(
+                                    stringExpr, (NewClassTree) tree, factory.getChecker());
+        } else {
+            throw new BugInCF("Unexpected tree: %s kind: %s", tree, tree.getKind());
+        }
+        convertAnnotatedTypeMirror(stringToJavaExpr, viewpointAdaptedType);
         this.viewpointAdaptedCopier.visit(viewpointAdaptedType, methodType);
     }
 
     /**
-     * Converts method or constructor arguments from Trees to JavaExpressions, accounting for
-     * varargs.
+     * Viewpoint-adapts the Java expressions in annotations written on a field declaration for the
+     * use at {@code fieldAccess}.
      *
-     * @param tree invocation of the method or constructor
-     * @param methodType type of the method or constructor
-     * @param argTrees the arguments to the method or constructor; subexpressions of {@code tree}
-     * @return the arguments, as JavaExpressions
+     * @param fieldAccess a field access
+     * @param type its type; is side-effected by this method
      */
-    private List<JavaExpression> argumentTreesToJavaExpressions(
-            ExpressionTree tree,
-            AnnotatedExecutableType methodType,
-            List<? extends ExpressionTree> argTrees) {
-
-        if (tree.getKind() == Kind.METHOD_INVOCATION) {
-            ExecutableElement method = TreeUtils.elementFromUse((MethodInvocationTree) tree);
-            if (isVarArgsInvocation(method, methodType, argTrees)) {
-                List<JavaExpression> result = new ArrayList<>();
-
-                for (int i = 0; i < method.getParameters().size() - 1; i++) {
-                    result.add(JavaExpression.fromTree(argTrees.get(i)));
-                }
-                List<JavaExpression> varargArgs = new ArrayList<>();
-                for (int i = method.getParameters().size() - 1; i < argTrees.size(); i++) {
-                    varargArgs.add(JavaExpression.fromTree(argTrees.get(i)));
-                }
-                Element varargsElement =
-                        method.getParameters().get(method.getParameters().size() - 1);
-                TypeMirror tm = ElementUtils.getType(varargsElement);
-                result.add(new ArrayCreation(tm, Collections.emptyList(), varargArgs));
-
-                return result;
-            }
+    public void atFieldAccess(MemberSelectTree fieldAccess, AnnotatedTypeMirror type) {
+        if (!hasDependentType(type) || TreeUtils.isClassLiteral(fieldAccess)) {
```

Comment:
Why is the special case needed?  (That is, can `hasDependentType` return true for a class literal?)
If it is needed, would it be more efficient to check it first?